sqlglot/optimizer/canonicalize.py

```diff
@@ -138,7 +138,7 @@ def _coerce_date(a: exp.Expression, b: exp.Expression) -> None:
             and b.type
             and b.type.this in exp.DataType.TEXT_TYPES
         ):
-            _replace_cast(b, exp.DataType.Type.DATETIME)
+            _replace_cast(b, a.type)
```

Comment:
That's a fair concern, I think we do... for MySQL:

```
mysql> SELECT CAST('2024-01-01' AS DATE) < '2024-01-01 12:00:00';
+----------------------------------------------------+
| CAST('2024-01-01' AS DATE) < '2024-01-01 12:00:00' |
+----------------------------------------------------+
|                                                  1 |
+----------------------------------------------------+
1 row in set (0.00 sec)
﻿```

But not for Postgres?!

```
georgesittas=# SELECT '2024-01-01'::DATE < '2024-01-01 12:00:00';
 ?column?
----------
 f
(1 row)
```

Reply:
Oy vey

Reply:
this canoncicalization is really only necessary for presto i think.

Reply:
Yeah, that’s why I added it. But it’s got to match the behavior of one or the other (or both, configurably)

Reply:
Another thing to think about: I believe cast(‘2024-01-01 00:00:01’) would fail in presto

Reply:
Should we keep it Presto/Trino-specific then by controlling it with a dialect argument? FWIW, for the above check:

- Databricks, DuckDB, Postgres, Redshift, Snowflake, Spark, T-SQL - return false
- MySQL - return true
- BigQuery, ClickHouse, Presto, Trino - fails

So this looks like MySQL weirdness? Based on these observations, I'd assume the others promote string -> whatever datetime type of the other operand.

Reply:
presto doesn't allow x < '2020-01-01' which is why it was added in the first place

Reply:
yes, i'd keep it specific to presto/trino.

Reply:
So, double-checking: keep the old logic but only have it kick in for Presto/Trino?

Reply:
Should the dialect arg be the parsed dialect or the target dialect?

I feel like the dialect arg should be avoided when possible, since it violates the idea that our AST is a superset of dialects

Reply:
it's the parser dialect. there's no notion of a target dialect.

i guess given that, this code isn't valid because x < text is not legal in presto

Reply:
another option is to just disable this for the ones that fail on type mismatch

Reply:
Right but we use the optimizer to convert MySQL queries to Trino where normal transpilation typically fails. I thought that was the point of canonicalize in the first place?

Reply:
if the goal is to parse mysql queries, you need to respect it,

thus x < '2022-01-01 01:00:00' should return false

Reply:
FYI I decided to investigate this a little further to make sure I understand MySQL's semantics well before moving on.

Reply:
I checked this [SO thread](https://stackoverflow.com/a/47561792) which helped, it seems like we're interested in this doc. [page](https://dev.mysql.com/doc/refman/8.4/en/date-and-time-type-conversion.html) in MySQL:

> Conversion of [DATE](https://dev.mysql.com/doc/refman/8.4/en/datetime.html) values:
>
>    Conversion to a [DATETIME](https://dev.mysql.com/doc/refman/8.4/en/datetime.html) or [TIMESTAMP](https://dev.mysql.com/doc/refman/8.4/en/datetime.html) value adds a time part of '00:00:00' because the [DATE](https://dev.mysql.com/doc/refman/8.4/en/datetime.html) value contains no time information.
>
>    Conversion to a [TIME](https://dev.mysql.com/doc/refman/8.4/en/time.html) value is not useful; the result is '00:00:00'.